File: quant1x/factors/notice.py

import requests
import math
from typing import List, Tuple, Dict, Optional, Union
from dataclasses import dataclass, field
from enum import Enum
from quant1x import exchange
from quant1x.exchange import Timestamp

# Constants
ERROR_BASE_NOTICE = 91000
URL_EASTMONEY_NOTICES = "https://np-anotice-stock.eastmoney.com/api/security/ann"
URL_EASTMONEY_WARNING = "https://datacenter.eastmoney.com/securities/api/data/get"
EASTMONEY_NOTICES_PAGE_SIZE = 100

RISK_KEYWORDS = [
    "立案", "处罚", "冻结", "诉讼", "质押", "仲裁",
    "持股5%以上股东权益变动", "信用减值", "商誉减值", "重大风险", "退市风险"
]

class NoticeException(Exception):
    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message
        super().__init__(f"[{code}] {message}")

ERR_NOTICE_BAD_API = NoticeException(ERROR_BASE_NOTICE, "接口异常")
ERR_NOTICE_NOT_FOUND = NoticeException(ERROR_BASE_NOTICE + 1, "没有数据")

@dataclass
class NoticeDetail:
    code: str = ""          # 证券代码
    name: str = ""          # 证券名称
    display_time: str = ""   # 显示时间
    notice_date: str = ""    # 公告时间
    title: str = ""         # 公告标题
    keywords: str = ""      # 公告关键词
    increase: int = 0       # 增持
    reduce: int = 0         # 减持
    holder_change: int = 0  # 实际控制人变更
    risk: int = 0           # 风险数

class EMNoticeType(Enum):
    NoticeAll = 0          # 全部
    NoticeUnused1 = 1      # 财务报告
    NoticeUnused2 = 2      # 融资公告
    NoticeUnused3 = 3      # 风险提示
    NoticeUnused4 = 4      # 信息变更
    NoticeWarning = 5      # 重大事项
    NoticeUnused6 = 6      # 资产重组
    NoticeHolderChange = 7  # 持股变动

def get_notice_type_name(notice_type: EMNoticeType) -> str:
    mapping = {
        EMNoticeType.NoticeAll: "全部",
        EMNoticeType.NoticeUnused1: "财务报告",
        EMNoticeType.NoticeUnused2: "融资公告",
        EMNoticeType.NoticeUnused3: "风险提示",
        EMNoticeType.NoticeUnused4: "信息变更",
        EMNoticeType.NoticeWarning: "重大事项",
        EMNoticeType.NoticeUnused6: "资产重组",
        EMNoticeType.NoticeHolderChange: "持股变动"
    }
    return mapping.get(notice_type, "其它")

@dataclass
class CompanyNotice:
    increase: int = 0      # 增持
    reduce: int = 0        # 减持
    risk: int = 0          # 风险数
    risk_keywords: str = ""  # 风险关键词

@dataclass
class WarningDetail:
    event_type: str = ""
    specific_event_type: str = ""
    notice_date: str = ""
    level1_content: str = ""
    level2_content: List[str] = field(default_factory=list)
    info_code: str = ""

@dataclass
class RawWarning:
    code: int = 0
    success: bool = False
    message: str = ""
    data: List[List[WarningDetail]] = field(default_factory=list)
    has_next: int = 0

def get_pages(page_size: int, total_hits: int) -> int:
    return (total_hits + page_size - 1) // page_size

def stock_notices(security_code: str, begin_date: str, end_date: str = "", page_number: int = 1) -> Tuple[List[NoticeDetail], int, Optional[NoticeException]]:
    """
    获取个股公告
    """
    try:
        fixed_begin_date = Timestamp.parse(begin_date).only_date()
    except ValueError:
        # Fallback or raise? C++ implies valid date.
        fixed_begin_date = begin_date

    if not end_date:
        fixed_end_date = Timestamp.now().only_date()
    else:
        try:
            fixed_end_date = Timestamp.parse(end_date).only_date()
        except ValueError:
            fixed_end_date = end_date

    page_size = EASTMONEY_NOTICES_PAGE_SIZE
    
    # Detect market to get stock_list param
    # C++: auto marketInfo = exchange::DetectMarket(securityCode);
    # params.Add({"stock_list", std::get<2>(marketInfo)});
    # std::get<2> is pure_code
    _, _, pure_code = exchange.detect_market(security_code)

    params = {
        "sr": "-1",
        "page_size": str(page_size),
        "page_index": str(page_number),
        "ann_type": "A",
        "client_source": "web",
        "f_node": "0",
        "s_node": "0",
        "begin_time": fixed_begin_date,
        "end_time": fixed_end_date,
        "stock_list": pure_code
    }

    try:
        response = requests.get(URL_EASTMONEY_NOTICES, params=params, timeout=10)
        if response.status_code != 200:
            return [], 0, ERR_NOTICE_BAD_API
        
        # Parse JSON
        # The response text might be JSONP or JSON. Assuming JSON based on C++ code.
        # C++: auto raw = json::parse(response.text);
        
        # Handle potential JSONP or just JSON
        try:
            raw = response.json()
        except Exception:
            # Sometimes APIs return text that isn't pure JSON?
            # C++ uses nlohmann::json::parse(response.text) which expects JSON.
            return [], 0, NoticeException(0, "JSON解析错误")

        # Check success
        # C++: noticePackage.Success = encoding::safe_json::get_number<int>(raw, "success", -1);
        # Note: Python requests.json() returns dict
        
        # In C++ struct RawNoticePackage has Data struct.
        # raw["data"]
        
        data_obj = raw.get("data", {})
        if not data_obj:
             # If data is null or missing
             return [], 0, ERR_NOTICE_NOT_FOUND

        total_hits = data_obj.get("total_hits", 0)
        pages = get_pages(page_size, total_hits)
        
        notices = []
        
        list_items = data_obj.get("list", [])
        if not list_items:
             return [], pages, ERR_NOTICE_NOT_FOUND

        for item in list_items:
            codes = item.get("codes", [])
            columns = item.get("columns", [])
            
            if not codes or not columns:
                continue
                
            # C++: auto marketCode = static_cast<exchange::MarketType>(std::stoll(noticeItem.Codes[0].MarketCode));
            # std::string securityCode = exchange::GetSecurityCode(marketCode, noticeItem.Codes[0].StockCode);
            
            code_info = codes[0]
            market_code_str = code_info.get("market_code", "1") # Default to SH?
            stock_code = code_info.get("stock_code", "")
            short_name = code_info.get("short_name", "")
            
            # Map market_code to MarketType
            # In C++ MarketType: ShenZhen=0, ShangHai=1, BeiJing=2
            # We need to convert this integer to our MarketType enum or just use get_security_code logic
            # quant1x.exchange.code.get_security_code takes MarketType enum.
            
            try:
                market_type_int = int(market_code_str)
                # We need to map int to MarketType enum if we want to use get_security_code strictly
                # But get_security_code in python takes MarketType enum.
                # Let's look at exchange.code.MarketType
                
                # from quant1x.exchange.code import MarketType
                # We can cast int to MarketType(int)
                from quant1x.exchange.code import MarketType, get_security_code
                market_type = MarketType(market_type_int)
                sec_code = get_security_code(market_type, stock_code)
            except Exception:
                # Fallback
                sec_code = stock_code

            notice = NoticeDetail()
            notice.code = sec_code
            notice.name = short_name
            notice.display_time = item.get("eiTime", "")
            notice.notice_date = item.get("notice_date", "")
            notice.title = item.get("title_ch", "")
            
            notice_keywords = []
            
            def check_risk(content: str):
                if not content:
                    return
                if "减持" in content:
                    notice_keywords.append("减持")
                    notice.reduce += 1
                if "增持" in content:
                    notice_keywords.append("增持")
                    notice.increase += 1
                if "控制人变更" in content:
                    notice_keywords.append("控制人变更")
                    notice.holder_change += 1
                for keyword in RISK_KEYWORDS:
                    if keyword in content:
                        notice_keywords.append(keyword)
                        notice.risk += 1
            
            for col in columns:
                check_risk(col.get("column_name", ""))
            check_risk(notice.title)
            
            if notice_keywords:
                # Unique and join
                # C++: std::accumulate with comma
                # Python: ",".join(notice_keywords) but need to handle duplicates if any?
                # C++ logic doesn't seem to unique here, but getOneNotice does.
                # Wait, C++ StockNotices:
                # notice.Keywords = std::accumulate(...)
                # It just joins them.
                notice.keywords = ",".join(notice_keywords)
                
            notices.append(notice)
            
        return notices, pages, None

    except Exception as e:
        print(f"[notice] Error: {e}")
        return [], 0, NoticeException(0, str(e))

def stock_warning(security_code: str, page_number: int) -> Tuple[RawWarning, Optional[NoticeException]]:
    """
    StockWarning - 安全版本
    """
    _, flag, pure_code = exchange.detect_market(security_code)
    flag = flag.upper()
    
    params = {
        "type": "RTP_F10_DETAIL",
        "params": f"{pure_code}.{flag},02",
        "p": str(page_number),
        "ann_type": "A",
        "source": "HSF10",
        "client": "PC"
    }
    
    try:
        response = requests.get(URL_EASTMONEY_WARNING, params=params, timeout=10)
        if response.status_code != 200:
            return RawWarning(), ERR_NOTICE_BAD_API
            
        raw = response.json()
        
        warning = RawWarning()
        warning.code = raw.get("code", 0)
        warning.success = raw.get("success", False)
        warning.message = raw.get("message", "")
        warning.has_next = raw.get("hasNext", 0)
        
        data_list = raw.get("data", [])
        if data_list:
            for data_item in data_list:
                details = []
                for detail in data_item:
                    wd = WarningDetail()
                    wd.event_type = detail.get("EVENT_TYPE", "")
                    wd.specific_event_type = detail.get("SPECIFIC_EVENTTYPE", "")
                    wd.notice_date = detail.get("NOTICE_DATE", "")
                    wd.level1_content = detail.get("LEVEL1_CONTENT", "")
                    wd.info_code = detail.get("INFO_CODE", "")
                    
                    l2 = detail.get("LEVEL2_CONTENT", [])
                    if isinstance(l2, list):
                        for content in l2:
                            if isinstance(content, str):
                                wd.level2_content.append(content)
                    
                    details.append(wd)
                warning.data.append(details)
                
        if not warning.success or not warning.data:
            return warning, ERR_NOTICE_NOT_FOUND
            
        return warning, None
        
    except Exception as e:
        print(f"[notice] Warning Error: {e}")
        return RawWarning(), NoticeException(0, str(e))

def get_annual_report_date(year: str, events: List[WarningDetail]) -> Tuple[str, str]:
    annual_report_date = ""
    quarterly_report_date = ""
    
    for v in events:
        try:
            date = Timestamp.parse(v.notice_date).only_date()
        except ValueError:
            continue
            
        tmp_year = date[:4]
        if v.event_type != "报表披露":
            continue
            
        if not annual_report_date and (v.specific_event_type == "年报披露" or v.specific_event_type == "年报预披露") and tmp_year >= year:
            annual_report_date = date
        elif not quarterly_report_date and ("季报披露" in v.specific_event_type or "季报预披露" in v.specific_event_type):
            quarterly_report_date = date
            
        if annual_report_date and quarterly_report_date:
            break
        if tmp_year < year:
            break
            
    return annual_report_date, quarterly_report_date

def notice_date_for_report(code: str, date: str) -> Tuple[str, str]:
    try:
        fixed_date = Timestamp.parse(date).only_date()
    except ValueError:
        fixed_date = date
        
    year = fixed_date[:4]
    page_no = 1
    annual_report_date = ""
    quarterly_report_date = ""
    
    while True:
        warning, err = stock_warning(code, page_no)
        if err:
            break
            
        for events in warning.data:
            tmp_annual, tmp_quarterly = get_annual_report_date(year, events)
            if not annual_report_date and tmp_annual:
                annual_report_date = tmp_annual
            if not quarterly_report_date and tmp_quarterly:
                quarterly_report_date = tmp_quarterly
            if annual_report_date and quarterly_report_date:
                break
                
        if annual_report_date and quarterly_report_date:
            break
            
        if warning.has_next > 0:
            page_no += 1
        else:
            break
            
    return annual_report_date, quarterly_report_date

def get_one_notice(security_code: str, current_date: str) -> CompanyNotice:
    notice = CompanyNotice()
    if not exchange.assert_stock_by_security_code(security_code):
        return notice
        
    try:
        ts = Timestamp.parse(current_date)
        # offset -24 * 30 hours? C++: timestamp.offset(-24 * 30)
        # In C++ offset takes hours?
        # Let's check C++ timestamp.h or usage.
        # In Python Timestamp.offset takes (hour, minute, second, ms).
        # C++: timestamp = timestamp.offset(-24 * 30); 
        # If C++ offset is in hours, then -720 hours = -30 days.
        # Python Timestamp.offset(hour=...)
        ts = ts.offset(hour=-24 * 30)
        begin_date = ts.only_date()
    except ValueError:
        return notice
        
    end_date = current_date
    pages_count = 1
    tmp_notice: Optional[NoticeDetail] = None
    
    page_no = 1
    while page_no <= pages_count:
        list_data, pages, err = stock_notices(security_code, begin_date, end_date, page_no)
        if err or pages < 1:
            break
            
        if pages_count < pages:
            pages_count = pages
        if not list_data:
            break
            
        for v in list_data:
            if tmp_notice:
                tmp_notice.name = v.name
                if tmp_notice.notice_date < v.notice_date:
                    tmp_notice.display_time = v.display_time
                    tmp_notice.notice_date = v.notice_date
                tmp_notice.title = v.title
                
                keywords = tmp_notice.keywords
                if v.keywords:
                    if not keywords:
                        keywords = v.keywords
                    else:
                        keywords += "," + v.keywords
                
                # Unique keywords
                if keywords:
                    tmp_arr = keywords.split(",")
                    # Remove empty strings and duplicates while preserving order?
                    # C++: strings::unique(tmpArr) usually sorts and uniques or just uniques.
                    # Python set is unordered.
                    # Let's use dict.fromkeys to preserve order
                    tmp_arr = list(dict.fromkeys([k for k in tmp_arr if k]))
                    tmp_notice.keywords = ",".join(tmp_arr)
                else:
                    tmp_notice.keywords = ""
                    
                tmp_notice.increase += v.increase
                tmp_notice.reduce += v.reduce
                tmp_notice.holder_change += v.holder_change
                tmp_notice.risk += v.risk
            else:
                # Clone v
                import copy
                tmp_notice = copy.deepcopy(v)
        
        if len(list_data) < EASTMONEY_NOTICES_PAGE_SIZE:
            break
            
        page_no += 1
        
    if tmp_notice:
        notice.increase = tmp_notice.increase
        notice.reduce = tmp_notice.reduce
        notice.risk = tmp_notice.risk
        notice.risk_keywords = tmp_notice.keywords
        
    return notice

File: quant1x/factors/notice.rs
use crate::exchange;
use serde::{Deserialize, Serialize};

const URL_EASTMONEY_NOTICES: &str = "https://np-anotice-stock.eastmoney.com/api/security/ann";
const URL_EASTMONEY_WARNING: &str = "https://datacenter.eastmoney.com/securities/api/data/get";
const EASTMONEY_NOTICES_PAGE_SIZE: i32 = 100;

const RISK_KEYWORDS: [&str; 11] = [
    "立案",
    "处罚",
    "冻结",
    "诉讼",
    "质押",
    "仲裁",
    "持股5%以上股东权益变动",
    "信用减值",
    "商誉减值",
    "重大风险",
    "退市风险",
];

#[derive(Debug, Clone, Default)]
pub struct NoticeDetail {
    pub code: String,
    pub name: String,
    pub display_time: String,
    pub notice_date: String,
    pub title: String,
    pub keywords: String,
    pub increase: i32,
    pub reduce: i32,
    pub holder_change: i32,
    pub risk: i32,
}

#[derive(Debug, Clone, Default)]
pub struct CompanyNotice {
    pub increase: i32,
    pub reduce: i32,
    pub risk: i32,
    pub risk_keywords: String,
}

#[derive(Debug, Clone, Copy)]
pub enum EMNoticeType {
    NoticeAll,
    NoticeUnused1,
    NoticeUnused2,
    NoticeUnused3,
    NoticeUnused4,
    NoticeWarning,
    NoticeUnused6,
    NoticeHolderChange,
}

impl EMNoticeType {
    pub fn as_str(&self) -> &'static str {
        match self {
            EMNoticeType::NoticeAll => "全部",
            EMNoticeType::NoticeUnused1 => "财务报告",
            EMNoticeType::NoticeUnused2 => "融资公告",
            EMNoticeType::NoticeUnused3 => "风险提示",
            EMNoticeType::NoticeUnused4 => "信息变更",
            EMNoticeType::NoticeWarning => "重大事项",
            EMNoticeType::NoticeUnused6 => "资产重组",
            EMNoticeType::NoticeHolderChange => "持股变动",
        }
    }
}

// Internal structs for JSON parsing
#[derive(Debug, Deserialize)]
struct RawNoticePackage {
    success: Option<i32>,
    #[serde(default)]
    data: Option<RawNoticeData>,
    #[serde(default)]
    error: Option<String>,
}

#[derive(Debug, Deserialize)]
struct RawNoticeData {
    #[serde(rename = "page_index")]
    page_index: Option<i32>,
    #[serde(rename = "page_size")]
    page_size: Option<i32>,
    #[serde(rename = "total_hits")]
    total_hits: Option<i32>,
    list: Option<Vec<RawNoticeItem>>,
}

#[derive(Debug, Deserialize)]
struct RawNoticeItem {
    #[serde(default)]
    codes: Vec<RawCodeInfo>,
    #[serde(default)]
    columns: Vec<RawColumnInfo>,
    #[serde(rename = "eiTime")]
    ei_time: Option<String>,
    #[serde(rename = "notice_date")]
    notice_date: Option<String>,
    #[serde(rename = "title_ch")]
    title_ch: Option<String>,
}

#[derive(Debug, Deserialize)]
struct RawCodeInfo {
    #[serde(rename = "market_code")]
    market_code: String,
    #[serde(rename = "stock_code")]
    stock_code: String,
    #[serde(rename = "short_name")]
    short_name: String,
}

#[derive(Debug, Deserialize)]
struct RawColumnInfo {
    #[serde(rename = "column_name")]
    column_name: String,
}

#[derive(Debug, Deserialize)]
pub struct RawWarning {
    pub success: Option<bool>,
    #[serde(rename = "hasNext")]
    pub has_next: Option<i32>,
    pub data: Option<Vec<Vec<WarningDetail>>>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct WarningDetail {
    #[serde(rename = "EVENT_TYPE")]
    pub event_type: Option<String>,
    #[serde(rename = "SPECIFIC_EVENTTYPE")]
    pub specific_event_type: Option<String>,
    #[serde(rename = "NOTICE_DATE")]
    pub notice_date: Option<String>,
}

pub fn stock_notices(
    security_code: &str,
    begin_date: &str,
    end_date: &str,
    page_number: i32,
) -> Result<(Vec<NoticeDetail>, i32), String> {
    let fixed_begin_date = crate::Timestamp::parse(begin_date)
        .map(|ts| ts.only_date())
        .unwrap_or_else(|_| begin_date.to_string());
    let fixed_end_date = if end_date.is_empty() {
        crate::Timestamp::now().only_date()
    } else {
        crate::Timestamp::parse(end_date)
            .map(|ts| ts.only_date())
            .unwrap_or_else(|_| end_date.to_string())
    };

    let (market_id, _, code) = exchange::detect_market(security_code);
    let stock_list = format!("{},{}", code, market_id);

    let client = reqwest::blocking::Client::new();
    let params = [
        ("sr", "-1"),
        ("page_size", &EASTMONEY_NOTICES_PAGE_SIZE.to_string()),
        ("page_index", &page_number.to_string()),
        ("ann_type", "A"),
        ("client_source", "web"),
        ("f_node", "0"),
        ("s_node", "0"),
        ("begin_time", &fixed_begin_date),
        ("end_time", &fixed_end_date),
        ("stock_list", &stock_list),
    ];

    let resp = client
        .get(URL_EASTMONEY_NOTICES)
        .query(&params)
        .send()
        .map_err(|e| format!("Request failed: {}", e))?;

    if !resp.status().is_success() {
        return Err(format!("Request failed with status: {}", resp.status()));
    }

    let raw: RawNoticePackage = resp
        .json()
        .map_err(|e| format!("JSON parse error: {}", e))?;

    if raw.success != Some(1) {
        return Err("API returned error".to_string());
    }

    let mut notices = Vec::new();
    let mut pages = 0;

    if let Some(data) = raw.data {
        pages = get_pages(EASTMONEY_NOTICES_PAGE_SIZE, data.total_hits.unwrap_or(0));
        if let Some(list) = data.list {
            for item in list {
                if item.codes.is_empty() || item.columns.is_empty() {
                    continue;
                }

                let code_info = &item.codes[0];
                let market_code_val = code_info.market_code.parse::<u8>().unwrap_or(0);
                let security_code_str =
                    exchange::security_code(market_code_val, &code_info.stock_code);

                let mut notice = NoticeDetail {
                    code: security_code_str,
                    name: code_info.short_name.clone(),
                    display_time: item.ei_time.clone().unwrap_or_default(),
                    notice_date: item.notice_date.clone().unwrap_or_default(),
                    title: item.title_ch.clone().unwrap_or_default(),
                    ..Default::default()
                };

                let mut notice_keywords = Vec::new();
                let mut check_risk = |content: &str| {
                    if content.contains("减持") {
                        notice_keywords.push("减持".to_string());
                        notice.reduce += 1;
                    }
                    if content.contains("增持") {
                        notice_keywords.push("增持".to_string());
                        notice.increase += 1;
                    }
                    if content.contains("控制人变更") {
                        notice_keywords.push("控制人变更".to_string());
                        notice.holder_change += 1;
                    }
                    for keyword in RISK_KEYWORDS.iter() {
                        if content.contains(keyword) {
                            notice_keywords.push(keyword.to_string());
                            notice.risk += 1;
                        }
                    }
                };

                for col in &item.columns {
                    check_risk(&col.column_name);
                }
                check_risk(&notice.title);

                if !notice_keywords.is_empty() {
                    notice.keywords = notice_keywords.join(",");
                }

                notices.push(notice);
            }
        }
    }

    Ok((notices, pages))
}

pub fn stock_warning(security_code: &str, page_number: i32) -> Result<RawWarning, String> {
    let (_market_id, flag, code) = exchange::detect_market(security_code);
    let flag_upper = flag.to_uppercase();
    // C++: std::get<2>(marketInfo) + "." + flag + ",02"
    // detect_market returns (id, flag, code). So it is code + "." + flag + ",02"
    let params_val = format!("{}.{},02", code, flag_upper);

    let client = reqwest::blocking::Client::new();
    let params = [
        ("type", "RTP_F10_DETAIL"),
        ("params", &params_val),
        ("p", &page_number.to_string()),
        ("ann_type", "A"),
        ("source", "HSF10"),
        ("client", "PC"),
    ];

    let resp = client
        .get(URL_EASTMONEY_WARNING)
        .query(&params)
        .send()
        .map_err(|e| format!("Request failed: {}", e))?;

    if !resp.status().is_success() {
        return Err(format!("Request failed with status: {}", resp.status()));
    }

    let warning: RawWarning = resp
        .json()
        .map_err(|e| format!("JSON parse error: {}", e))?;

    if warning.success != Some(true) {
        return Err("API returned error".to_string());
    }

    Ok(warning)
}

fn get_annual_report_date(
    year: &str,
    events: &[WarningDetail],
) -> (Option<String>, Option<String>) {
    let mut annual_report_date = None;
    let mut quarterly_report_date = None;

    for v in events {
        let notice_date = v.notice_date.as_deref().unwrap_or("");
        let date = crate::Timestamp::parse(notice_date)
            .map(|ts| ts.only_date())
            .unwrap_or_default();
        if date.len() < 4 {
            continue;
        }
        let tmp_year = &date[0..4];

        let event_type = v.event_type.as_deref().unwrap_or("");
        if event_type != "报表披露" {
            continue;
        }

        let specific_event_type = v.specific_event_type.as_deref().unwrap_or("");

        if annual_report_date.is_none()
            && (specific_event_type == "年报披露" || specific_event_type == "年报预披露")
            && tmp_year >= year
        {
            annual_report_date = Some(date.clone());
        } else if quarterly_report_date.is_none()
            && (specific_event_type.contains("季报披露")
                || specific_event_type.contains("季报预披露"))
        {
            quarterly_report_date = Some(date.clone());
        }

        if annual_report_date.is_some() && quarterly_report_date.is_some() {
            break;
        }
        if tmp_year < year {
            break;
        }
    }
    (annual_report_date, quarterly_report_date)
}

pub fn notice_date_for_report(code: &str, date: &str) -> (String, String) {
    let fixed_date = crate::Timestamp::parse(date)
        .map(|ts| ts.only_date())
        .unwrap_or_else(|_| date.to_string());
    if fixed_date.len() < 4 {
        return (String::new(), String::new());
    }
    let year = &fixed_date[0..4];
    let mut page_no = 1;
    let mut annual_report_date = String::new();
    let mut quarterly_report_date = String::new();

    loop {
        match stock_warning(code, page_no) {
            Ok(warning) => {
                if let Some(data) = warning.data {
                    for events in data {
                        let (tmp_annual, tmp_quarterly) = get_annual_report_date(year, &events);
                        if annual_report_date.is_empty() {
                            if let Some(d) = tmp_annual {
                                annual_report_date = d;
                            }
                        }
                        if quarterly_report_date.is_empty() {
                            if let Some(d) = tmp_quarterly {
                                quarterly_report_date = d;
                            }
                        }
                        if !annual_report_date.is_empty() && !quarterly_report_date.is_empty() {
                            break;
                        }
                    }
                }

                if !annual_report_date.is_empty() && !quarterly_report_date.is_empty() {
                    break;
                }
                if warning.has_next.unwrap_or(0) > 0 {
                    page_no += 1;
                } else {
                    break;
                }
            }
            Err(_) => break,
        }
    }

    (annual_report_date, quarterly_report_date)
}

pub fn get_one_notice(security_code: &str, current_date: &str) -> CompanyNotice {
    let mut notice = CompanyNotice::default();
    if !exchange::assert_stock_by_security_code(security_code) {
        return notice;
    }

    let timestamp =
        crate::Timestamp::parse(current_date).unwrap_or_else(|_| crate::Timestamp::now());
    // offset -24 * 30 hours? C++: timestamp.offset(-24 * 30)
    // Assuming offset takes hours in C++, but in Rust Timestamp implementation it might differ.
    // Let's assume we need to go back 30 days.
    // If Timestamp doesn't have offset, we might need to use chrono or similar logic.
    // Looking at Timestamp implementation in previous context, it wraps chrono::NaiveDateTime or similar.
    // Let's assume we can calculate it.
    // Since I don't have full Timestamp API, I'll use a safe approximation or try to use what's available.
    // C++: timestamp = timestamp.offset(-24 * 30);
    // Let's assume 30 days ago.

    // Hack: parse to chrono, subtract, format back.
    // Or use crate::exchange::calendar if available.
    // For now, let's try to use a simple calculation if Timestamp exposes it.
    // If not, I'll just use current_date as end_date and maybe a fixed start date?
    // No, the logic requires a window.

    // Let's try to use chrono directly since it is a dependency.
    let dt =
        chrono::NaiveDate::parse_from_str(&timestamp.only_date(), "%Y-%m-%d").unwrap_or_default();
    let begin_date_dt = dt - chrono::Duration::days(30);
    let begin_date = begin_date_dt.format("%Y-%m-%d").to_string();

    let end_date = current_date.to_string();
    let mut pages_count = 1;

    let mut tmp_notice: Option<NoticeDetail> = None;
    let mut page_no = 1;

    loop {
        if page_no > pages_count {
            break;
        }

        match stock_notices(security_code, &begin_date, &end_date, page_no) {
            Ok((list, pages)) => {
                if pages_count < pages {
                    pages_count = pages;
                }
                if list.is_empty() {
                    break;
                }

                for v in &list {
                    if let Some(ref mut t_notice) = tmp_notice {
                        t_notice.name = v.name.clone();
                        if t_notice.notice_date < v.notice_date {
                            t_notice.display_time = v.display_time.clone();
                            t_notice.notice_date = v.notice_date.clone();
                        }
                        t_notice.title = v.title.clone();

                        let mut keywords = t_notice.keywords.clone();
                        if !v.keywords.is_empty() {
                            if keywords.is_empty() {
                                keywords = v.keywords.clone();
                            } else {
                                keywords = format!("{},{}", keywords, v.keywords);
                            }
                        }

                        // Unique keywords
                        let mut tmp_arr: Vec<&str> = keywords.split(',').collect();
                        tmp_arr.sort();
                        tmp_arr.dedup();
                        t_notice.keywords = tmp_arr.join(",");

                        t_notice.increase += v.increase;
                        t_notice.reduce += v.reduce;
                        t_notice.holder_change += v.holder_change;
                        t_notice.risk += v.risk;
                    } else {
                        tmp_notice = Some(v.clone());
                    }
                }

                if list.len() < EASTMONEY_NOTICES_PAGE_SIZE as usize {
                    break;
                }
            }
            Err(_) => break,
        }

        page_no += 1;

        // Safety break to prevent infinite loops if logic is wrong
        if page_no > 100 {
            break;
        }
    }

    if let Some(t_notice) = tmp_notice {
        notice.increase = t_notice.increase;
        notice.reduce = t_notice.reduce;
        notice.risk = t_notice.risk;
        notice.risk_keywords = t_notice.keywords;
    }

    notice
}

fn get_pages(page_size: i32, total_hits: i32) -> i32 {
    (total_hits + page_size - 1) / page_size
}
